src/replication/adapter/awsecr/adapter.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsecr
+
+import (
+	"errors"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awsecrapi "github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/common/utils/registry"
+	adp "github.com/goharbor/harbor/src/replication/adapter"
+	"github.com/goharbor/harbor/src/replication/model"
+	"net/http"
+	"regexp"
+)
+
+func init() {
+	if err := adp.RegisterFactory(model.RegistryTypeAwsEcr, func(registry *model.Registry) (adp.Adapter, error) {
+		return newAdapter(registry)
+	}); err != nil {
+		log.Errorf("failed to register factory for %s: %v", model.RegistryTypeAwsEcr, err)
+		return
+	}
+	log.Infof("the factory for adapter %s registered", model.RegistryTypeAwsEcr)
+}
+
+func newAdapter(registry *model.Registry) (*adapter, error) {
+	region, err := parseRegion(registry.URL)
+	if err != nil {
+		return nil, err
+	}
+	authorizer := NewAuth(region, registry.Credential.AccessKey, registry.Credential.AccessSecret, registry.Insecure)
+	reg, err := adp.NewDefaultImageRegistryWithCustomizedAuthorizer(registry, authorizer)
+	if err != nil {
+		return nil, err
+	}
+	return &adapter{
+		registry:             registry,
+		DefaultImageRegistry: reg,
+	}, nil
+}
+
+func parseRegion(url string) (string, error) {
+	pattern := "https://(?:api|\\d+\\.dkr)\\.ecr\\.([\\w\\-]+)\\.amazonaws\\.com"
+	rs := regexp.MustCompile(pattern).FindStringSubmatch(url)
+	if rs == nil {
+		return "", errors.New("Bad aws url")
+	}
+	return rs[1], nil
+}
+
+type adapter struct {
+	*adp.DefaultImageRegistry
+	registry *model.Registry
+}
+
+var _ adp.Adapter = adapter{}
+
+func (adapter) Info() (info *model.RegistryInfo, err error) {
```

Comment:
Please use pointer receiver and keep the same with other adapter implements.

Reply:
ok
actually 'src/replication/adapter/native/adapter.go' also have the same problem.